File: bind.c
/* bind.c */
/*****************************************************************************/
/* AS-Portierung                                                             */
/*                                                                           */
/* Bearbeitung von AS-P-Dateien                                              */
/*                                                                           */
/* Historie:  1. 6.1996 Grundsteinlegung                                     */
/*                                                                           */
/*****************************************************************************/

#include "stdinc.h"
#include <ctype.h>
#include <string.h>

#include "endian.h"
#include "stdhandl.h"
#include "bpemu.h"
#include "stringutil.h"
#include "decodecmd.h"
#include "asmutils.h"
#include "nls.h"

static char *Creator="BIND/C 1.41r5";


static CMDProcessed ParProcessed;
static Integer z;

static FILE *TargFile;
static String TargName;

#include "tools.rsc"
#include "bind.rsc"

	static void OpenTarget(void)
BEGIN
   TargFile=fopen(TargName,"w");
   if (TargFile==Nil) ChkIO(TargName);
   if (NOT Write2(TargFile,&FileID)) ChkIO(TargName); 
END

	static void CloseTarget(void)
BEGIN
   Byte EndHeader=FileHeaderEnd;

   if (fwrite(&EndHeader,1,1,TargFile)!=1) ChkIO(TargName);
   if (fwrite(Creator,1,strlen(Creator),TargFile)!=strlen(Creator)) ChkIO(TargName);
   if (fclose(TargFile)==EOF) ChkIO(TargName);
   if (Magic!=0) unlink(TargName);
END

	static void ProcessFile(char *FileName)
BEGIN
#define BufferSize 8192
   FILE *SrcFile;
   Word TestID;
   Byte InpHeader,InpSegment,InpGran;
   LongInt InpStart,SumLen;
   Word InpLen,TransLen;
   Boolean doit;
   Byte Buffer[BufferSize];

   SrcFile=fopen(FileName,"r");
   if (SrcFile==Nil) ChkIO(FileName);

   if (NOT Read2(SrcFile,&TestID)) ChkIO(FileName);
   if (TestID!=FileMagic) FormatError(FileName,FormatInvHeaderMsg);

   errno=0; printf("%s==>>%s",FileName,TargName); ChkIO(OutName);

   SumLen=0;

   do
    BEGIN
     ReadRecordHeader(&InpHeader,&InpSegment,&InpGran,FileName,SrcFile);
     if (InpHeader==FileHeaderStartAdr)
      BEGIN
       if (NOT Read4(SrcFile,&InpStart)) ChkIO(FileName);
       WriteRecordHeader(&InpHeader,&InpSegment,&InpGran,TargName,TargFile);
       if (NOT Write4(TargFile,&InpStart)) ChkIO(TargName);
      END
     else if (InpHeader!=FileHeaderEnd)
      BEGIN
       if (NOT Read4(SrcFile,&InpStart)) ChkIO(FileName);
       if (NOT Read2(SrcFile,&InpLen)) ChkIO(FileName);

       if (ftell(SrcFile)+InpLen>=FileSize(SrcFile)-1)
        FormatError(FileName,FormatInvRecordLenMsg);

       doit=FilterOK(InpHeader);

       if (doit)
        BEGIN
	 SumLen+=InpLen;
         WriteRecordHeader(&InpHeader,&InpSegment,&InpGran,TargName,TargFile);
	 if (NOT Write4(TargFile,&InpStart)) ChkIO(TargName);
	 if (NOT Write2(TargFile,&InpLen)) ChkIO(TargName);
	 while (InpLen>0)
	  BEGIN
           TransLen=min(BufferSize,InpLen);
	   if (fread(Buffer,1,TransLen,SrcFile)!=TransLen) ChkIO(FileName);
	   if (fwrite(Buffer,1,TransLen,TargFile)!=TransLen) ChkIO(TargName);
	   InpLen-=TransLen;
	  END
        END
       else
        BEGIN
         if (fseek(SrcFile,InpLen,SEEK_CUR)==-1) ChkIO(FileName);
        END
      END
    END 
   while (InpHeader!=FileHeaderEnd);

   errno=0; printf("  (%d Byte",SumLen); ChkIO(OutName);

   if (fclose(SrcFile)==EOF) ChkIO(FileName);
END

	static void ProcessGroup(char *GroupName)
BEGIN/**
   s:SearchRec;
   Path,Name,Ext:String;**/
   String Name;

   strmaxcpy(Name,GroupName,255);
   AddSuffix(Name,Suffix);
   ProcessFile(Name);
/**
   FSplit(GroupName,Path,Name,Ext);

   FindFirst(GroupName,Archive,s);
   WHILE DosError=0 DO
    BEGIN
     ProcessFile(Path+s.Name);
     FindNext(s);
    END;**/
END

	static void ParamError(Boolean InEnv, char *Arg)
BEGIN
   printf("%s%s\n",InEnv ? ErrMsgInvEnvParam : ErrMsgInvParam,Arg);
   printf("%s\n",ErrMsgProgTerm);
   exit(1);
END

#define BINDParamCnt 1
static CMDRec BINDParams[BINDParamCnt]=
	      {{"f", CMD_FilterList}};

	void main(int argc, char **argv)
BEGIN
   ParamCount=argc-1; ParamStr=argv;

   NLS_Initialize(); WrCopyRight("BIND V1.41r5");

   stdhandl_init(); decodecmd_init(); asmutils_init(); nls_init();
   if (ParamCount==0)
    BEGIN
     errno=0; printf("%s%s%s\n",InfoMessHead1,GetEXEName(),InfoMessHead2); ChkIO(OutName);
     for (z=0; z<InfoMessHelpCnt; z++)
      BEGIN
       errno=0; printf("%s\n",InfoMessHelp[z]); ChkIO(OutName);
      END
     exit(1);
    END

   ProcessCMD(BINDParams,BINDParamCnt,ParProcessed,"BINDCMD",ParamError);

   z=ParamCount;
   while ((z>0) AND (NOT ParProcessed[z])) z--;
   if (z==0)
    BEGIN
     errno=0; printf("%s\n",ErrMsgTargetMissing);
     ChkIO(OutName);
     exit(1);
    END
   else
    BEGIN
     strmaxcpy(TargName,ParamStr[z],255); ParProcessed[z]=False;
     AddSuffix(TargName,Suffix);
    END

   OpenTarget();

   for (z=1; z<=ParamCount; z++)
    if (ParProcessed[z]) ProcessGroup(ParamStr[z]);

   CloseTarget();
END

File: bind.pas
{$g-,i-,v-}
	PROGRAM Bind;
        Uses Dos,StdHandl,DecodeCm,AsmUtils,NLS;

CONST
   {$IFDEF OS2}
   Creator:ARRAY[1..14] OF Char='BIND/2 1.41r7';
   {$ELSE}
   Creator:ARRAY[1..12] OF Char='BIND 1.41r7';
   {$ENDIF}

VAR
   ParProcessed:CMDProcessed;
   z,z2:Integer;

   TargFile:File;
   TargName:String;

{$i Tools.rsc}
{$i Bind.rsc}

	PROCEDURE OpenTarget;
BEGIN
   Assign(TargFile,TargName); SetFileMode(2); Rewrite(TargFile,1);
   ChkIO(TargName);
   BlockWrite(TargFile,FileID,2);
   ChkIO(TargName);
END;

	PROCEDURE CloseTarget;
CONST
   EndHeader:Byte=FileHeaderEnd;
BEGIN
   BlockWrite(TargFile,EndHeader,1); ChkIO(TargName);
   BlockWrite(TargFile,Creator,Sizeof(Creator)); ChkIO(TargName);
   Close(TargFile); ChkIO(TargName);
   IF Magic<>0 THEN Erase(TargFile);
END;

	PROCEDURE ProcessFile(FileName:String);
CONST
   BufferSize=8192;
VAR
   SrcFile:File;
   TestID:Word;
   InpHeader,InpSegment,InpGran:Byte;
   InpStart,SumLen:LongInt;
   InpLen,TransLen:Word;
   doit:Boolean;
   Buffer:ARRAY[1..BufferSize] OF Byte;
   z:Integer;
BEGIN
   Assign(SrcFile,FileName); SetFileMode(0); Reset(SrcFile,1);
   ChkIO(FileName);

   BlockRead(SrcFile,TestID,2);
   IF TestID<>FileMagic THEN FormatError(FileName,FormatInvHeaderMsg);

   Write(FileName,'==>>',TargName); ChkIO(OutName);

   SumLen:=0;

   REPEAT
    ReadRecordHeader(InpHeader,InpSegment,InpGran,FileName,SrcFile);
    IF InpHeader=FileHeaderStartAdr THEN
     BEGIN
      BlockRead(SrcFile,InpStart,4); ChkIO(FileName);
      WriteRecordHeader(InpHeader,InpSegment,InpGran,TargName,TargFile);
      BlockWrite(TargFile,InpStart,4); ChkIO(TargName);
     END
    ELSE IF InpHeader<>FileHeaderEnd THEN
     BEGIN
      BlockRead(SrcFile,InpStart,4); ChkIO(FileName);
      BlockRead(SrcFile,InpLen,2); ChkIO(FileName);

      IF FilePos(SrcFile)+InpLen>=FileSize(SrcFile)-1 THEN
       FormatError(FileName,FormatInvRecordLenMsg);

      doit:=FilterOK(InpHeader);

      IF doit THEN
       BEGIN
	Inc(SumLen,InpLen);
        WriteRecordHeader(InpHeader,InpSegment,InpGran,TargName,TargFile);
	BlockWrite(TargFile,InpStart,4); ChkIO(TargName);
	BlockWrite(TargFile,InpLen,2); ChkIO(TargName);
	WHILE InpLen>0 DO
	 BEGIN
	  IF InpLen>BufferSize THEN TransLen:=BufferSize ELSE TransLen:=InpLen;
	  BlockRead(SrcFile,Buffer,TransLen); ChkIO(FileName);
	  BlockWrite(TargFile,Buffer,TransLen); ChkIO(TargName);
	  Dec(InpLen,TransLen);
	 END;
       END
      ELSE
       BEGIN
        Seek(SrcFile,FilePos(SrcFile)+InpLen); ChkIO(FileName);
       END;
     END;
   UNTIL InpHeader=0;

   WriteLn('  (',SumLen,' Byte)'); ChkIO(OutName);

   Close(SrcFile); ChkIO(FileName);
END;

	PROCEDURE ProcessGroup(GroupName:String);
VAR
   s:SearchRec;
   Path,Name,Ext:String;
BEGIN
   AddSuffix(GroupName,Suffix);

   FSplit(GroupName,Path,Name,Ext);

   FindFirst(GroupName,Archive,s);
   WHILE DosError=0 DO
    BEGIN
     ProcessFile(Path+s.Name);
     FindNext(s);
    END;
END;

	PROCEDURE ParamError(InEnv:Boolean; Arg:String);
	Far;
BEGIN
   IF InEnv THEN Write(ErrMsgInvEnvParam)
	    ELSE Write(ErrMsgInvParam);
   WriteLn(Arg);
   WriteLn(ErrMsgProgTerm);
   Halt(1);
END;

CONST
   BINDParamCnt=1;
   BINDParams:ARRAY[1..BINDParamCnt] OF CMDRec=
	      ((Ident:'f'; Callback:CMD_FilterList));

BEGIN
   NLS_Initialize; WrCopyRight('BIND/2 V1.41r7','BIND V1.41r7');

   IF ParamCount=0 THEN
    BEGIN
     WriteLn(InfoMessHead1,GetEXEName,InfoMessHead2); ChkIO(OutName);
     FOR z:=1 TO InfoMessHelpCnt DO
      BEGIN
       WriteLn(InfoMessHelp[z]); ChkIO(OutName);
      END;
     Halt(1);
    END;

   ProcessCMD(@BINDParams,BINDParamCnt,ParProcessed,'BINDCMD',ParamError);

   z:=ParamCount;
   WHILE (z>0) AND (NOT (z IN ParProcessed)) DO Dec(z);
   IF z=0 THEN
    BEGIN
     WriteLn(ErrMsgTargetMissing);
     ChkIO(OutName);
     Halt(1);
    END
   ELSE
    BEGIN
     TargName:=ParamStr(z); ParProcessed:=ParProcessed-[z];
     AddSuffix(TargName,Suffix);
    END;

   OpenTarget;

   FOR z:=1 TO ParamCount DO
    IF z IN ParProcessed THEN ProcessGroup(ParamStr(z));

   CloseTarget;
END.
